common/net/conn.go

```diff
@@ -0,0 +1,205 @@
+package net
+
+import (
+	"encoding/binary"
+	"fmt"
+	"math"
+	"net"
+)
+
+const (
+	// MsgLenFieldSize is the byte size of the frame length field of a
+	// framed message.
+	MsgLenFieldSize = 4
+	// The byte size of the message type field of a framed message.
+	msgTypeFieldSize = 4
+	// The bytes size limit for a record message.
+	recordLengthLimit = 1024 * 1024 // 1 MiB
+	// The default bytes size of a record message.
+	recordDefaultLength = 4 * 1024 // 4KiB
+	// Message type value included in record framing.
+	recordMsgType = uint32(0x06)
+	// The initial write buffer size.
+	writeBufferInitialSize = 32 * 1024 // 32KiB
+	// The maximum write buffer size. This *must* be multiple of
+	// recordDefaultLength.
+	writeBufferMaxSize = 512 * 1024 // 512KiB
+)
+
+// conn represents a secured connection. It implements the net.Conn interface.
+type conn struct {
+	net.Conn
+	crypto Cipher
+	// buf holds data that has been read from the connection and decrypted,
+	// but has not yet been returned by Read.
+	buf                []byte
+	payloadLengthLimit int
+	// protected holds data read from the network but have not yet been
+	// decrypted. This data might not compose a complete frame.
+	protected []byte
+	// writeBuf is a buffer used to contain encrypted frames before being
+	// written to the network.
+	writeBuf []byte
+	// nextFrame stores the next frame (in protected buffer) info.
+	nextFrame []byte
+	// overhead is the calculated overhead of each frame.
+	overhead int
+}
+
+// NewConn creates a new secure channel instance given the other party role and
+// handshaking result.
+func NewConn(c net.Conn, handshaker Handshaker) (net.Conn, error) {
```

Comment:
```suggestion
func NewConn(c net.Conn, side Side, handshaker Handshaker) (net.Conn, error) {
```
Remove `IsClient` from `Handshaker`. `Handshaker` will not know on which side it's working on.

Reply:
thx